# Importing essential libraries

In [1]:
import selenium
import time
from selenium import webdriver

from bs4 import BeautifulSoup
import pandas as pd

# Loading the `Webdriver`

In [2]:
driver = webdriver.Chrome()

# Getting the access of `ACT Plumber's` official website

In [38]:
driver.get('https://www.accesscanberra.act.gov.au/s/public-registers/construction-licences?registerid=m0I4a00000000CMEAY')
time.sleep(2)

# Finding number of pages of the data 

In [39]:
pages=driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/p').text
pages

'Page: 1 / 215'

In [40]:
numbers = []
for word in pages.split():
    if word.isdigit():
        numbers.append(int(word))
print(numbers[1])

215


# Main program to get `Headers` & `all the data rows` of just the `first page`

In [41]:
page_source = driver.page_source
soup =  BeautifulSoup(page_source, 'lxml')
            
data = [j.text for j in soup.findAll('td')]
data
all_names= []
all_names= data[::5]
all_names

['360 DEGREE FIRE PTY LTD',
 'A & A HOTWATER & GAS SOLUTIONS PTY LTD',
 'A & K (ACT) PTY LTD',
 'A & K (ACT) PTY LTD',
 'A PLUS PLUMBING (ACT) PTY LTD',
 'A PLUS PLUMBING (ACT) PTY LTD',
 'AAA ABOVE BOARD PLUMBING PTY LTD',
 'AAA ABOVE BOARD PLUMBING PTY LTD',
 'ABBEY, DAMIEN LETTON',
 'ABDEL RAHMAN, MUDATHIR NOURIN ALI']

In [42]:
data= data[3:]
#data
all_class_desc= []
all_class_desc= data[::5]
all_class_desc

['Water Supply Plumber',
 'Water Supply Plumber',
 'Sanitary Plumber',
 'Water Supply Plumber',
 'Water Supply Plumber',
 'Sanitary Plumber',
 'Water Supply Plumber',
 'Sanitary Plumber',
 'Sanitary Plumber',
 'Journeyperson Plumber']

# Generating xpaths for all the records in any page

In [43]:
xpaths=[]
for j in range(1,11): #creating xpaths: just change span[i] to click different links
    xpaths.append('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr['+str(j) + ']/td[5]/lightning-button/button')
print(xpaths) 

['/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[1]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[2]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[3]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[4]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[5]/td[5]/lightning-button/button', '/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/di

In [44]:
licensee_detail=[]
for k in range(1,11): #getting the details of every license in a combination
    driver.find_element_by_xpath(xpaths[k-1]).click()
    #time.sleep(1)
    
    page_source = driver.page_source
    soup = BeautifulSoup(page_source, 'lxml')
            
    det= [j.text for j in soup.findAll('td')]
    det=det[1:]
    #det

    ab=[]
    ab=det[::2]

    ab.insert(0,all_names[k-1])
    ab.insert(3,all_class_desc[k-1])
    ab

    licensee_detail.append(ab)
    
    
    
    
    
    driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div/lightning-button/button').click()

In [45]:
licensee_detail

[['360 DEGREE FIRE PTY LTD',
  '2017990',
  'Plumber',
  'Water Supply Plumber',
  '22 January 2023',
  ''],
 ['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
  '20191321',
  'Plumber',
  'Water Supply Plumber',
  '2 February 2024',
  ''],
 ['A & K (ACT) PTY LTD',
  '2016855',
  'Plumber',
  'Sanitary Plumber',
  '13 September 2022',
  '',
  ''],
 ['A & K (ACT) PTY LTD',
  '2016855',
  'Plumber',
  'Water Supply Plumber',
  '13 September 2022',
  '',
  ''],
 ['A PLUS PLUMBING (ACT) PTY LTD',
  '2016683',
  'Plumber',
  'Water Supply Plumber',
  '17 August 2022',
  '',
  ''],
 ['A PLUS PLUMBING (ACT) PTY LTD',
  '2016683',
  'Plumber',
  'Sanitary Plumber',
  '17 August 2022',
  '',
  ''],
 ['AAA ABOVE BOARD PLUMBING PTY LTD',
  '2018299',
  'Plumber',
  'Water Supply Plumber',
  '31 May 2022',
  '',
  ''],
 ['AAA ABOVE BOARD PLUMBING PTY LTD',
  '2018299',
  'Plumber',
  'Sanitary Plumber',
  '31 May 2022',
  '',
  ''],
 ['ABBEY, DAMIEN LETTON',
  '200216119',
  'Plumber',
  'Sanitary Plumbe

In [46]:
licensee_detail[1]

['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
 '20191321',
 'Plumber',
 'Water Supply Plumber',
 '2 February 2024',
 '']

In [47]:
len(licensee_detail)

10

# Main program to get `Headers` & `all the data rows` of all the pages till 215

In [48]:
for i in range(1, numbers[1]):
    
    driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/div[4]/button').click()
    time.sleep(1)
    
    page_source = driver.page_source
    soup =  BeautifulSoup(page_source, 'lxml')       
    data = [j.text for j in soup.findAll('td')]
    #data
    all_names= []
    all_names= data[::5]
    #all_names
    data= data[3:]
    #data
    all_class_desc= []
    all_class_desc= data[::5]
    #all_class_desc
    
    
    for k in range(1,11): 
        
        driver.find_element_by_xpath(xpaths[k-1]).click()
        #time.sleep(1)
    
        page_source = driver.page_source
        soup = BeautifulSoup(page_source, 'lxml')
            
        det= [l.text for l in soup.findAll('td')]
        det=det[1:]
        #det

        ab=[]
        ab=det[::2]

        ab.insert(0,all_class_desc[k-1])
        ab.insert(0,all_names[k-1])
        #ab

        licensee_detail.append(ab)
    
        driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div/lightning-button/button').click()

NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[5]/div/div/table/tbody/tr[8]/td[5]/lightning-button/button"}
  (Session info: chrome=93.0.4577.82)


In [51]:
det

['201583',
 'Occupation',
 'Plumber',
 'Expiry date',
 '1 June 2024',
 'Water Supply Plumber',
 '',
 'Sanitary Plumber',
 '']

In [52]:
ab

['ZOUCH, GEORGE ANDREW',
 'Sanitary Plumber',
 '201583',
 'Plumber',
 '1 June 2024',
 '',
 '']

In [53]:
licensee_detail

[['360 DEGREE FIRE PTY LTD',
  '2017990',
  'Plumber',
  'Water Supply Plumber',
  '22 January 2023',
  ''],
 ['A & A HOTWATER & GAS SOLUTIONS PTY LTD',
  '20191321',
  'Plumber',
  'Water Supply Plumber',
  '2 February 2024',
  ''],
 ['A & K (ACT) PTY LTD',
  '2016855',
  'Plumber',
  'Sanitary Plumber',
  '13 September 2022',
  '',
  ''],
 ['A & K (ACT) PTY LTD',
  '2016855',
  'Plumber',
  'Water Supply Plumber',
  '13 September 2022',
  '',
  ''],
 ['A PLUS PLUMBING (ACT) PTY LTD',
  '2016683',
  'Plumber',
  'Water Supply Plumber',
  '17 August 2022',
  '',
  ''],
 ['A PLUS PLUMBING (ACT) PTY LTD',
  '2016683',
  'Plumber',
  'Sanitary Plumber',
  '17 August 2022',
  '',
  ''],
 ['AAA ABOVE BOARD PLUMBING PTY LTD',
  '2018299',
  'Plumber',
  'Water Supply Plumber',
  '31 May 2022',
  '',
  ''],
 ['AAA ABOVE BOARD PLUMBING PTY LTD',
  '2018299',
  'Plumber',
  'Sanitary Plumber',
  '31 May 2022',
  '',
  ''],
 ['ABBEY, DAMIEN LETTON',
  '200216119',
  'Plumber',
  'Sanitary Plumbe

In [60]:
licensee_detail[0]

['360 DEGREE FIRE PTY LTD',
 '2017990',
 'Plumber',
 'Water Supply Plumber',
 '22 January 2023',
 '']

In [61]:
len(licensee_detail)

2147

# Creating the final dataframe by zipping all the attribute's lists

In [62]:
df = pd.DataFrame(licensee_detail, columns = ['Name', 'License number','Occupation','Class description','Expiry Date','Class Condition / Endorsement','x1','x2'])
 
# print dataframe.
df

Name  \
0                           360 DEGREE FIRE PTY LTD   
1            A & A HOTWATER & GAS SOLUTIONS PTY LTD   
2                               A & K (ACT) PTY LTD   
3                               A & K (ACT) PTY LTD   
4                     A PLUS PLUMBING (ACT) PTY LTD   
5                     A PLUS PLUMBING (ACT) PTY LTD   
6                  AAA ABOVE BOARD PLUMBING PTY LTD   
7                  AAA ABOVE BOARD PLUMBING PTY LTD   
8                              ABBEY, DAMIEN LETTON   
9                 ABDEL RAHMAN, MUDATHIR NOURIN ALI   
10                                    ABSOLOM, ALEX   
11                         ACKLAND, CHRISTOPHER LEE   
12                         ACKLAND, CHRISTOPHER LEE   
13                                     ADAMS, JAMES   
14                                ADENS-KANTS, MARK   
15                                ADENS-KANTS, MARK   
16    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD   
17    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD   
18                                 AGGARWAL, SURESH   
19                                 AGGARWAL, SURESH   
20                                   AGIUS, MICHAEL   
21                                   AGIUS, MICHAEL   
22                           AGOSTINO, ALEX DOMINIC   
23                           AGOSTINO, ALEX DOMINIC   
24                                  AL-KULABI, SAIF   
25                                  AL-KULABI, SAIF   
26                                AL-ZAHROUT, BRAKH   
27                 ALL GAS PLUMBING REPAIRS PTY LTD   
28                 ALL GAS PLUMBING REPAIRS PTY LTD   
29                              ALLAN, JAMES GORDON   
...                                             ...   
2117                        WOODWARD, ADAM KINGSLEY   
2118                        WOODWARD, ADAM KINGSLEY   
2119                          WOODWARD, MARK VICTOR   
2120                          WOODWARD, MARK VICTOR   
2121                             WRIGHT, ALLAN JOHN   
2122                             WRIGHT, ALLAN JOHN   
2123                                    WRIGHT, BEN   
2124                                    WRIGHT, BEN   
2125                          WRIGHT, JAMES BERNARD   
2126                                  WRIGHT, OSCAR   
2127                         WRIGHT, WILLIAM JOSEPH   
2128                         WRIGHT, WILLIAM JOSEPH   
2129                        WROE, CARRINGTON ROBERT   
2130                       WUBBELS, BRANDON MICHAEL   
2131                                  WYBORN, ETHAN   
2132                                  WYBORN, ETHAN   
2133                                   WYBURN, SEAN   
2134                                   WYBURN, SEAN   
2135                            XL PLUMBING PTY LTD   
2136                            XL PLUMBING PTY LTD   
2137                                 YOUNG, MICHAEL   
2138              YOUR LOCAL PLUMBING GROUP PTY LTD   
2139                                 ZAMMIT, EDWARD   
2140                                 ZAMMIT, EDWARD   
2141                                   ZHANG, CHENG   
2142                            ZINGLEMANN, GREGORY   
2143                            ZINGLEMANN, GREGORY   
2144                             ZORZI, LUKE SAMUEL   
2145                           ZOUCH, GEORGE ANDREW   
2146                           ZOUCH, GEORGE ANDREW   

                           License number Occupation      Class description  \
0                                 2017990    Plumber   Water Supply Plumber   
1                                20191321    Plumber   Water Supply Plumber   
2                                 2016855    Plumber       Sanitary Plumber   
3                                 2016855    Plumber   Water Supply Plumber   
4                                 2016683    Plumber   Water Supply Plumber   
5                                 2016683    Plumber       Sanitary Plumber   
6                                 2018299    Plumber   Water Supply Plumber   
7   

In [65]:
df = df.drop(['x1','x2'], 1)

In [67]:
df

Name  \
0                           360 DEGREE FIRE PTY LTD   
1            A & A HOTWATER & GAS SOLUTIONS PTY LTD   
2                               A & K (ACT) PTY LTD   
3                               A & K (ACT) PTY LTD   
4                     A PLUS PLUMBING (ACT) PTY LTD   
5                     A PLUS PLUMBING (ACT) PTY LTD   
6                  AAA ABOVE BOARD PLUMBING PTY LTD   
7                  AAA ABOVE BOARD PLUMBING PTY LTD   
8                              ABBEY, DAMIEN LETTON   
9                 ABDEL RAHMAN, MUDATHIR NOURIN ALI   
10                                    ABSOLOM, ALEX   
11                         ACKLAND, CHRISTOPHER LEE   
12                         ACKLAND, CHRISTOPHER LEE   
13                                     ADAMS, JAMES   
14                                ADENS-KANTS, MARK   
15                                ADENS-KANTS, MARK   
16    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD   
17    ADVANCED PLUMBING & DRAINS (CANBERRA) PTY LTD   
18                                 AGGARWAL, SURESH   
19                                 AGGARWAL, SURESH   
20                                   AGIUS, MICHAEL   
21                                   AGIUS, MICHAEL   
22                           AGOSTINO, ALEX DOMINIC   
23                           AGOSTINO, ALEX DOMINIC   
24                                  AL-KULABI, SAIF   
25                                  AL-KULABI, SAIF   
26                                AL-ZAHROUT, BRAKH   
27                 ALL GAS PLUMBING REPAIRS PTY LTD   
28                 ALL GAS PLUMBING REPAIRS PTY LTD   
29                              ALLAN, JAMES GORDON   
...                                             ...   
2117                        WOODWARD, ADAM KINGSLEY   
2118                        WOODWARD, ADAM KINGSLEY   
2119                          WOODWARD, MARK VICTOR   
2120                          WOODWARD, MARK VICTOR   
2121                             WRIGHT, ALLAN JOHN   
2122                             WRIGHT, ALLAN JOHN   
2123                                    WRIGHT, BEN   
2124                                    WRIGHT, BEN   
2125                          WRIGHT, JAMES BERNARD   
2126                                  WRIGHT, OSCAR   
2127                         WRIGHT, WILLIAM JOSEPH   
2128                         WRIGHT, WILLIAM JOSEPH   
2129                        WROE, CARRINGTON ROBERT   
2130                       WUBBELS, BRANDON MICHAEL   
2131                                  WYBORN, ETHAN   
2132                                  WYBORN, ETHAN   
2133                                   WYBURN, SEAN   
2134                                   WYBURN, SEAN   
2135                            XL PLUMBING PTY LTD   
2136                            XL PLUMBING PTY LTD   
2137                                 YOUNG, MICHAEL   
2138              YOUR LOCAL PLUMBING GROUP PTY LTD   
2139                                 ZAMMIT, EDWARD   
2140                                 ZAMMIT, EDWARD   
2141                                   ZHANG, CHENG   
2142                            ZINGLEMANN, GREGORY   
2143                            ZINGLEMANN, GREGORY   
2144                             ZORZI, LUKE SAMUEL   
2145                           ZOUCH, GEORGE ANDREW   
2146                           ZOUCH, GEORGE ANDREW   

                           License number Occupation      Class description  \
0                                 2017990    Plumber   Water Supply Plumber   
1                                20191321    Plumber   Water Supply Plumber   
2                                 2016855    Plumber       Sanitary Plumber   
3                                 2016855    Plumber   Water Supply Plumber   
4                                 2016683    Plumber   Water Supply Plumber   
5                                 2016683    Plumber       Sanitary Plumber   
6                                 2018299    Plumber   Water Supply Plumber   
7   

## Checking the shape (rows & columns) of the final dataframe

In [69]:
df.shape

(2147, 6)

## Converting the dataframe into a `.csv` file

In [70]:
df.to_csv('ACT_Plumbers.csv')